### DISCUSSION TITLE
How to do something after a service responded to a request?

### DISCUSSION THREAD

**Discussion by johannescpk** [hyperium/hyper#2731] [Q&A]
If I have a `service_fn` that returns a response and I'd like to do something _after_ the response was fully sent, how would I approach that?

**Comment by johannescpk** [comment]
~~After looking around a bit, it seems the only viable solution is writing a tower service for the request handler and wrapping that into a [layer](https://docs.rs/tower/latest/tower/trait.Layer.html).~~

Nope, doesn't work either as that only transforms the request/response.

**Comment by olix0r** [comment]
I'd use `tokio::spawn` to create a background task that runs independently of the request's task.

**Reply by johannescpk** [reply]
Good idea, thanks. I suppose you don't have a guarantee that the spawned task actually runs after the request fully completed tho, right?

**Reply by olix0r** [reply]
Generally, that's a very hard problem to solve. If you really need to guarantee that some action happens, you'd have to write to some sort of durable work queue before responding so that if process crashes after the response is sent the work can still continue. Even if you're not trying to defend against that sort of catastrophe, you can't really _guarantee_ that an arbitrary background task will ever complete.

Or if you're trying to ensure that the task doesn't _start_ running until the response is completed: I would probably create a custom `Body` type that holds an `Arc<tokio::sync::Notify>` with a `Drop` impl that calls `Notify::notify_waiters`. The `tokio::spawn` task can await on `Notify::notified` to be notified when the response has been fully processed by the hyper server.
